### ISSUE TITLE
Add possibility to override lifetime for some messages (or by severity)

### ISSUE THREAD

**Issue by sebaiona** [kreuzerk/primeNG-advanced-growl#154]
Hello,

And first of all thank you for this great extension!

I'd like to know if there is a possibility to set a custom `lifetime` for some messages (or by severity).
For example , in my case I'd like to set errorMessages with a `life`=0 so that the user has to close it manually (by this way I'm sure they see it).

**Comment by kreuzerk** [comment]
Hy @sebaiona 

Thanks a lot for submitting this feature request! I think that this is a great idea! Yes it should be possible. 

I would suggest to extend the service methods createSuccessMessage, createInfoMessage etc... to accept a lifetime for each message that you create. With this extension you should then be able to achieve what you want to do. The goal would be to provide you with the following API

this.advMessagesService.createSuccessMessage('Awesome message', 'Awesome summery', lifetime);

I still need to check if the lifetime property then becomes obsolete.

In conclusion. It is possible and I think that this is a great idea. I will try to implement this in the next weeks. But I still need to figure out how the API should look like.

**Comment by sebaiona** [comment]
Thanks a lot !

I think it would be interesting to have both possibilities : 
- default lifetime
- optional lifetime for each message

By this way if you don't set a lifetime on the call it would take the default lifetime.

For me it's an important feature so let me know when you do it and if you need tests.

**Comment by kreuzerk** [comment]
Hi @sebaiona 

I have published an alpha version with the desired feature to set the lifeTime individually for each created message to npm. It would be great if you have the possibility to test this version and let me know if it fits your need and if the API is understandable. Unfortunately I have not found the time yet to document the API and adjust the unit test. Here's a quick documentation on what you should do to test the alpha version:

First install it via:
`npm install --save primeng-advanced-growl@alpha
`

Then you can use the library as normal. To use the new feature just pass in the lifeTime as optional parameter to the service methods.

`        this.advMessagesService.createInfoMessage('Awesome success message content', 'Awesome info', 3000);`

If you want the message to be sticky (to stay on the screen) like you want it for the error message then pass 0 as parameter.

`        this.advMessagesService.createErrorMessage('Awful error message', 'Error summary', 0);`

Notice that if you don't pass the optional lifeTime parameter to the service methods it will take the default one initially passed to the component via input parameter.

If you have further questions we can discuss here or we can also chat on the chatroom (https://gitter.im/primeNG-advanced-growl/Lobby)

Looking forward to your feedback!

**Comment by sebaiona** [comment]
Hi @kreuzerk !

I just tried it quickly and it seems to work perfectly fine !
I have error messages with a lifetime=0 and other messages with the default one. 
Everything seems to have the behavior it supposed to have.

I had a call with additionalProperties so it didn't compile, but what you did is in my mind the best choice for the API.
It's easily understandable and perhaps there are bugs but I didn't find any.

Now every message is really independent and configurable !!
It fits exactly what I need and I think it's a great new feature.

Thanks a lot!

**Comment by kreuzerk** [comment]
@sebaiona Great!! Love to hear that it worked!

As you mentioned the call with additionalProperties will not compile anymore. This is not so bad because we use semver and we will then release version 4 and introduce this as a breaking change.

I am happy with the new API but there is still one thing I do not like. If somebody for example just wants their messages to have the same lifeTime and use additionalProperties he will no longer be able to just pass the additional Properties without a lifeTime. He then needs to pass the lifeTime in every place he uses additionalProperties. 

I will research if there is still a better way for the API. If not I will adjust the documentation, adjust the unit tests and prepare the final release for this version.

**Comment by sebaiona** [comment]
Indeed, the lifetime is needed if you want to call with additionalProperties (even if you can set it at `null`). 

A quick workaround would be to add `createInfoMessage`, `createErrorMessage`, .... with the same signatures as they were before (without lifetime) and make them call the main functions with a `lifetime = null`.

It's not really the kind of solution I like but it's transparent for the users of the API.

**Comment by kreuzerk** [comment]
@sebaiona I like the idea of a transparent API for the user. It would also be great because the changes would be non breaking. Generally I prefer non breaking changes due to easier update and migration for library consumers.

I am not a huge fan of providing a overloaded method createSuccessMessage where it is possible to either pass in the lifeTime and the additionalProperties or just the lifeTime. Overloading in TypeScript is a little bit different from other languages like Java or C#. Because I would need to manually check neither you passed in a lifeTime number or something else. 

I would prefer to have two Methods on the Service:
`createSuccessMessage(messageContent: string, summary: string, additionalProperties?: any): void `

and another
`createTimedSuccessMessage(messageContent: string, summary: string, lifeTime: number, additionalProperties?: any): void `

What do you think about this API?

**Comment by sebaiona** [comment]
Choosing a non breaking solution for the library consumers is clearly the best solution.
No need to change anything after the update is for the best.

Your solution with the two methods looks good to me. 
It will do the same thing but the user won't see it  👍 

Let me know if you want me to test a new version ;)

**Comment by kreuzerk** [comment]
@sebaiona Sorry that it took me so long to get back to you. I have just adjusted the unit tests and added the new service methods to create timed messages. You now have the normal methods createSuccessMessage etc... like today. Additionally you now also have the following methods:

- createTimedSuccessMessage(messageContent: string, summary: string, lifeTime: number, additionalProperties?: any): void
- createTimedInfoMessage(messageContent: string, summary: string, lifeTime: number, additionalProperties?: any): void
- createTimedErrorMessage(messageContent: string, summary: string, lifeTime: number, additionalProperties?: any): void
- createTimedWarningMessage(messageContent: string, summary: string, lifeTime: number, additionalProperties?: any): void

Can you please test if it still fits your needs. If a get a positiv feedback from you I will then go on and publish this version under 3.1.0. Initially documentation will be missing but it will follow. I just want to publish this version as quick as possible so that you and maybe other people who see the new methods can already use it.

Documentation will then follow step by step. First in the readme and then we will also have to rethink the demo page. 

Thanks for your help. Looking forward to your feedback!

**Comment by sebaiona** [comment]
Hi @kreuzerk,

I just tested your new methods and everything looks fine to me.
Method names seems pretty clear. It's easy to understand and there is no confusion with the already existing ones (at least in my mind).

I tried `createTimedWarningMessage` and `createTimedErrorMessage` with a `lifetime=0` (which I already tried in the alpha version) and it seems to be as functionnal as it was before !

Everything seems to be working great ! 👍
